EIPS/eip-3584.md

```diff
@@ -0,0 +1,131 @@
+```
+Set [ AccessedAddress, 
+      List [AccessedStorageKeys] , 
+      Set  [ AccessedInBlockTransactionNumber, List [ AccessedStorageKeys ]]  ]
+```
+
+To prevent any *grieving* attacks, an `AccessListRoot` : a fingerpint of a **canonical** specification of `access_list` will need to be included (and validated) in the *block header*.
+
+#### Canonical Block Access List
+An `access_list` is defined to be comprised of many `access_list_entry` elements:
+```
+access_list   :=  [access_list_entry, ...]
+```
+
+An `access_list_entry` is a 3-tuple of:
+* address
+* sorted list of storage keys of the address accessed across the entire block
+* sorted list of 2-tuples of:
+    * transaction index in which the address or any of its storage keys were accessed
+    * sorted list of storage keys which were accessed
+
+```
+access_list                 := [access_list_entry, ...]
+access_list_entry           := [address, storage_keys, accesses_by_txn_index]
+address                     := bytes20
+accesses_by_txn_index       := [txn_index_and_keys, ...]
+txn_index_and_keys          := [txn_index, storage_keys]
+txn_index                   := uint64  # or uint256 or whatever
+storage_keys                := [storage_key, ...]
+storage_key                 := bytes32
+```
+
+Additional sorting rules for the above are that:
+* `access_list` is sorted by the `address`
+* `storage_keys` is sorted
+* `accesses_by_txn_index` is sorted by `txn_index`
+
+Additional validation rules for the above are that:
+* Each unique `address` may only appear at most once in `access_list`
+* Each `storage_key` may only appear at most once in `storage_keys`
+* Each `txn_index` may only appear at most once in `txn_index_and_keys`
+
+All sorting is in increasing order.
+
+#### AccessListRoot
+An `AccessListRoot` is a [urn](https://en.wikipedia.org/wiki/Uniform_Resource_Name) encoding `Hash/Commitment` of the **canonical** `access_list` as well as the construction type (`sha256/kate/merkle/verkle`).
+
+For a very simple start we can do a `SHA256` of the canonical `access_list` and include it in the block header with the following the urn scheme:
+```
+AccessListRoot := "urn:sha256:${ SHA256( access_list ) }"
+```
+
+This AccessListRoot could evolve to `urn:merkle:...` or to `urn:kate:...` or to any other scheme as per requirement.
+
+#### Additional Block Validation
+Full nodes while accepting, executing the new block have an additional validation check that the block’s `AccessListRoot` matches the one calculated from the one generated by executing the block using the construction as defined by the `AccessListRoot` urn.
+
+#### A JSON RPC Endpoint
+Recommended implementation of fetching the block access list of a given block something like 
+```
+eth_BlockAccessList ( BlockNumberOrBlockHashOrAccessListURN, ListOfAddressesOrOffsetOrNull )
+``` 
+which can return paginated `access_list`, paginated at the top address level, filtered by optional addresses filter.
+
```

Comment:
This should not be included.  Core EIPs focus on consensus changes, while changes to JSON-RPC happen in another repository and usually *after* the core EIP is finalized.
```suggestion
```